README.md
![Ruby](https://github.com/rafaelsales/ulid/workflows/Ruby/badge.svg)
[![Gem Downloads](http://img.shields.io/gem/dt/ulid.svg)](https://rubygems.org/gems/ulid)
[![GitHub License](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/rafaelsales/ulid)

# ulid
Universally Unique Lexicographically Sortable Identifier implementation for Ruby

Official specification page: https://github.com/ulid/spec

<h1 align="center">
	<br>
	<br>
	<img width="360" src="logo.png" alt="ulid">
	<br>
	<br>
	<br>
</h1>

# Universally Unique Lexicographically Sortable Identifier

UUID can be suboptimal for many uses-cases because:

- It isn't the most character efficient way of encoding 128 bits of randomness
- The string format itself is apparently based on the original MAC & time version (UUIDv1 from Wikipedia)
- It provides no other information than randomness

Instead, herein is proposed ULID:

- 128-bit compatibility with UUID
- 1.21e+24 unique ULIDs per millisecond
- Lexicographically sortable!
- Canonically encoded as a 26 character string, as opposed to the 36 character UUID
- Uses Crockford's base32 for better efficiency and readability (5 bits per character)
- Case insensitive
- No special characters (URL safe)

### Installation

```
gem install ulid
```

### Usage

```ruby
require 'ulid'

ULID.generate # 01ARZ3NDEKTSV4RRFFQ69G5FAV
```

**I want to generate a ULID using an arbitrary timestamp**

You can optionally pass a `Time` instance to `ULID.generate` to set an arbitrary timestamp component, i.e. the prefix of the ULID.

```ruby
time_t1 = Time.now
ulid = ULID.generate(time_t1)
```

**I want to generate a ULID using an arbitrary suffix, i.e. without the randomness component**

You can optionally pass a 80-bit hex-encodable `String` on the argument `suffix` to `ULID.generate`. This will replace the randomness component
by the suffix provided. This allows for fully deterministic ULIDs.

```ruby
require 'securerandom'

time = Time.now
an_event_identifier = SecureRandom.uuid
ulid1 = ULID.generate(time, suffix: an_event_identifier)
ulid2 = ULID.generate(time, suffix: an_event_identifier)
ulid1 == ulid2 # true
```

## Specification

Below is the current specification of ULID as implemented in this repository. *Note: the binary format has not been implemented.*

```
 01AN4Z07BY      79KA1307SR9X4MV3

|----------|    |----------------|
 Timestamp          Randomness
  10 chars           16 chars
   48bits             80bits
   base32             base32
```

### Components

**Timestamp**
- 48 bit integer
- UNIX-time in milliseconds
- Won't run out of space till the year 10895 AD.

**Randomness**
- 80 bits
- Cryptographically secure source of randomness, if possible

### Sorting

The left-most character must be sorted first, and the right-most character sorted last. The default ASCII order is used for sorting.

### Binary Layout and Byte Order

The components are encoded as 16 octets. Each component is encoded with the Most Significant Byte first (network byte order).

```
0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                      32_bit_uint_time_high                    |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|     16_bit_uint_time_low      |       16_bit_uint_random      |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                       32_bit_uint_random                      |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                       32_bit_uint_random                      |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

### String Representation

```
ttttttttttrrrrrrrrrrrrrrrr

where
t is Timestamp
r is Randomness
```

## Test Suite

```
bundle exec rake test
```

### Credits and references:

* https://github.com/ulid/javascript


CHANGELOG.md
# 1.4.0

- PR #32 - Allow fully deterministic ULID generation by providing both the timestamp
  component and a string suffix to replace the randomness component. Thanks @andjosh

# 1.2.0

- PR #20 - Use an array to improve speed / reduce memory allocations. Thanks, @jamescook

# 1.1.1

- PR #19 - Remove Timecop gem. Thanks, @bquorning

- PR #18 - Remove Mocha gem. Thanks, @bquorning

- PR #17 - Add post_install_message to install sysrandom gem

- PR #16 - Remove circular require warning. Thanks, @pocke

# 1.1.0

- PR #14 - Ruby 2.5+ does not need sysrandom. Thanks, @sakuro

# 1.0.0

- PR #10 - Fix time encoding. Thanks, @dcuddeback


spec/spec_helper.rb
require 'byebug'
require 'minitest/autorun'
require 'minitest/pride'
require 'ulid'


spec/lib/ulid_spec.rb
require 'spec_helper'
require 'base32/crockford'

describe ULID do
  describe 'textual representation' do
    it 'ensures it has 26 chars' do
      ulid = ULID.generate

      assert_equal ulid.length, 26
    end

    it 'is sortable' do
      input_time = Time.now
      ulid1 = ULID.generate(input_time)
      ulid2 = ULID.generate(input_time + 1)
      assert ulid2 > ulid1
    end

    it 'is valid Crockford Base32' do
      ulid = ULID.generate
      decoded = Base32::Crockford.decode(ulid)
      encoded = Base32::Crockford.encode(decoded, length: 26)
      assert_equal ulid, encoded
    end

    it 'encodes the timestamp in the first 10 characters' do
      # test case taken from original ulid README:
      # https://github.com/ulid/javascript#seed-time
      #
      # N.b. we avoid specifying the time as a float, since we lose precision:
      #
      # > Time.at(1_469_918_176.385).strftime("%F %T.%N")
      # => "2016-07-30 23:36:16.384999990"
      #
      # vs the correct:
      #
      # > Time.at(1_469_918_176, 385, :millisecond).strftime("%F %T.%N")
      # => "2016-07-30 23:36:16.385000000"
      ulid = ULID.generate(Time.at(1_469_918_176, 385, :millisecond))
      assert_equal '01ARYZ6S41', ulid[0...10]
    end

    it 'respects millisecond-precision order' do
      ulids = Array.new(1000) do |millis|
        time = Time.new(2020, 1, 2, 3, 4, Rational(millis, 10**3))

        ULID.generate(time)
      end

      assert_equal(ulids, ulids.sort)
    end

    it 'is deterministic based on time' do
      input_time = Time.now
      ulid1 = ULID.generate(input_time)
      ulid2 = ULID.generate(input_time)
      assert_equal ulid2.slice(0, 10), ulid1.slice(0, 10)
      assert ulid2 != ulid1
    end

    it 'is deterministic based on suffix' do
      input_time = Time.now
      suffix = SecureRandom.uuid
      ulid1 = ULID.generate(input_time, suffix: suffix)
      ulid2 = ULID.generate(input_time + 1, suffix: suffix)
      assert_equal ulid2.slice(10, 26), ulid1.slice(10, 26)
      assert ulid2 != ulid1
    end

    it 'is fully deterministic based on time and suffix' do
      input_time = Time.now
      suffix = SecureRandom.uuid
      ulid1 = ULID.generate(input_time, suffix: suffix)
      ulid2 = ULID.generate(input_time, suffix: suffix)
      assert_equal ulid2, ulid1
    end

    it 'raises exception when non-encodable 80-bit suffix string is used' do
      input_time = Time.now
      suffix = SecureRandom.uuid
      assert_raises(ArgumentError) do
        ULID.generate(input_time, suffix: suffix[0...9])
      end

      ULID.generate(input_time, suffix: suffix[0...10])
    end
  end

  describe 'underlying binary' do
    it 'encodes the timestamp in the high 48 bits' do
      input_time = Time.now.utc
      bytes = ULID.generate_bytes(input_time)
      (time_ms,) = "\x0\x0#{bytes[0...6]}".unpack('Q>')
      encoded_time = Time.at(time_ms / 1000.0).utc
      assert_in_delta input_time, encoded_time, 0.001
    end

    it 'encodes the remaining 80 bits as random' do
      random_bytes = SecureRandom.random_bytes(ULID::Generator::RANDOM_BITS / 8)
      SecureRandom.stub(:random_bytes, random_bytes) do
        bytes = ULID.generate_bytes
        assert bytes[6..-1] == random_bytes
      end
    end
  end
end


lib/ulid/generator.rb
# frozen-string-literal: true

require 'securerandom'

module ULID
  module Generator
    ENCODING = '0123456789ABCDEFGHJKMNPQRSTVWXYZ'.bytes.freeze # Crockford's Base32
    RANDOM_BITS = 80
    ENCODED_LENGTH = 26
    BITS_PER_B32_CHAR = 5
    ZERO = '0'.ord
    MASK = 0x1f

    # Generates a 128-bit ULID string.
    # @param [Time] time (Time.now) Timestamp - first 48 bits
    # @param [String] suffix (80 random bits) - the remaining 80 bits as hex encodable string
    def generate(time = Time.now, suffix: nil)
      (hi, lo) = generate_bytes(time, suffix: suffix).unpack('Q>Q>')
      if hi.nil? || lo.nil?
        raise ArgumentError, 'suffix string without hex encoding passed to ULID generator'
      end

      integer = (hi << 64) | lo
      encode(integer, ENCODED_LENGTH)
    end

    # Generates a 128-bit ULID.
    # @param [Time] time (Time.now) Timestamp - first 48 bits
    # @param [String] suffix (80 random bits) - the remaining 80 bits as hex encodable string
    def generate_bytes(time = Time.now, suffix: nil)
      suffix_bytes =
        if suffix
          suffix.split('').map { |char| char.to_i(32) }.pack('C*')
        else
          SecureRandom.random_bytes(RANDOM_BITS / 8)
        end

      time_48bit(time) + suffix_bytes
    end

    private

    # Encodes a 128-bit integer input as a 26-character ULID string using Crockford's Base32.
    def encode(input, length)
      encoded = Array.new(length, ZERO)
      i = length - 1

      while input > 0
        encoded[i] = ENCODING[input & MASK]
        input >>= BITS_PER_B32_CHAR
        i -= 1
      end

      encoded.pack('c*')
    end

    # Returns the first 6 bytes of a timestamp (in milliseconds since the Unix epoch) as a 48-bit byte string
    def time_48bit(time = Time.now)
      # Avoid `time.to_f` since we want to accurately represent a whole number of milliseconds:
      #
      # > time = Time.new(2020, 1, 5, 7, 3, Rational(2, 1000))
      # => 2020-01-05 07:03:00 +0000
      # > (time.to_f * 1000).to_i
      # => 1578207780001
      #
      # vs
      #
      # > (time.to_r * 1000).to_i
      # => 1578207780002
      time_ms = (time.to_r * 1000).to_i
      [time_ms].pack('Q>')[2..-1]
    end
  end
end
